apps/src/blockly/addons/cdoAngleHelper.ts

```diff
@@ -0,0 +1,533 @@
+import * as Blockly from 'blockly/core';
+
+interface AngleHelperOptions {
+  arcColour?: string;
+  angle?: number;
+  height?: number;
+  width?: number;
+  snapPoints?: number[];
+  onUpdate?: () => void;
+  enableBackgroundRotation?: boolean;
+}
+
+class Vector {
+  constructor(public x: number, public y: number) {}
+
+  clone(): Vector {
+    return new Vector(this.x, this.y);
+  }
+
+  add(vec: Vector): Vector {
+    return new Vector(this.x + vec.x, this.y + vec.y);
+  }
+
+  subtract(vec: Vector): Vector {
+    return new Vector(this.x - vec.x, this.y - vec.y);
+  }
+
+  rotate(angle: number): Vector {
+    const rad = angle * (Math.PI / 180);
+    const cos = Math.cos(rad);
+    const sin = Math.sin(rad);
+    const newX = this.x * cos - this.y * sin;
+    const newY = this.y * cos + this.x * sin;
+    return new Vector(newX, newY);
+  }
+
+  static rotateAroundPoint(vec: Vector, center: Vector, angle: number): Vector {
+    const res = vec.clone().subtract(center).rotate(angle);
+    return res.add(center);
+  }
+
+  static distance(vec1: Vector, vec2: Vector): number {
+    return Math.sqrt(
+      Math.pow(vec2.x - vec1.x, 2) + Math.pow(vec2.y - vec1.y, 2)
+    );
+  }
+}
+
+class CdoAngleHelper {
+  private lineColour_: string;
+  private strokeWidth_: number;
+  private arcColour_: string;
+  private height_: number;
+  public width_: number;
+  private snapPoints_?: number[];
+  private onUpdate_?: () => void;
+  private enableBackgroundRotation_: boolean;
+  private turnRight_: boolean;
+  private picker_: {
+    handleRadius: number;
+    isDragging: boolean;
+    angle: number;
+    line: SVGLineElement | null;
+    handle: SVGCircleElement | null;
+    handleCenter?: Vector;
+  };
+  private background_: {
+    handleRadius: number;
+    tickSpacing: number;
+    isDragging: boolean;
+    angle: number;
+    handle: SVGCircleElement | null;
+    line: SVGLineElement | null;
+    ticks: SVGLineElement[];
+    handleCenter?: Vector;
+  };
+  private radius_: Vector;
+  private center_: Vector;
+  private arc_: SVGPathElement | null;
+  public svg_: SVGSVGElement | null;
+  private rect_: DOMRect | null;
+  private animationInterval_: number | undefined;
+  private mouseMoveWrapper_: ((e: MouseEvent) => void) | null;
+  private mouseUpWrapper_: ((e: MouseEvent) => void) | null;
+  private mouseDownWrapper_: ((e: MouseEvent) => void) | null;
+
+  constructor(direction: string, options: AngleHelperOptions = {}) {
+    this.lineColour_ = '#4d575f';
+    this.strokeWidth_ = 3;
```

Comment:
There are unfortunately a ton of little magic numbers throughout this file, most of which are only used once. Ideally, some of these values would be computed based on a set scale for the entire component. Since everything is static, I'm inclined to leave them as is for now, rather define a whole bunch of constants with super context-specifc names.

Reply:
Makes sense. I do see `'turnRight'` a few times, would that one be worth a constant?

Reply:
Yes, I've done that here: https://github.com/code-dot-org/code-dot-org/pull/59789/commits/de25569f59b3bb6188e9d7a7e2b87922c7ac73de#diff-2340c7f9b234c61841daecd90ec745870b3bab1c80d511f68b6b07dedfcac3b8R79

I'll double-check to make sure I didn't miss any instances!